### PR TITLE
build: use parallelization except on Windows

### PR DESCRIPTION
**Pull request by cathyjf**
This is similar to LizardByte/Sunshine#3361, except now multithreaded
processing is not used on Windows.

Unlike LizardByte/Sunshine#3361, we're now enabling parallelization
for both dot graph generation and other doxygen processing tasks as well
(except on Windows).

To check for Windows, we use the macro `CMAKE_HOST_WIN32` rather than
the more well-known `WIN32`, because what matters here is the host
platform, not the target platform. In most cases, these will be the
same, but the issue with parallelization is related to Windows as a
host, not a target. In other words, using Windows to build the
documentation for a macOS project should still be single-threaded.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [X] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -92,13 +92,23 @@ add_custom_command(
...
+endif()
+
 # build docs
 add_custom_target(docs ALL
         COMMENT "Building Doxygen documentation"
         WORKING_DIRECTORY "${SOURCE_DOCS_DIR}"
         COMMAND ${CMAKE_COMMAND} -E env
             READTHEDOCS_OUTPUT=${DOXYGEN_BUILD_DIR_RELATIVE}
             READTHEDOCS_VERSION=${DOXYGEN_PROJECT_VERSION}
+            READTHEDOCS_THREADS=${DOXYGEN_NUM_THREADS}
```
Comment (ReenigneArcher) [inline CMakeLists.txt:111]:
```suggestion
            DOXYCONFIG_THREADS=${DOXYGEN_NUM_THREADS}
```

The other two env variables are standard readthedocs variables: https://docs.readthedocs.io/en/stable/reference/environment-variables.html

File: doxyconfig-Doxyfile [hunk 2]
```diff
@@ -107,7 +107,7 @@ DISABLE_INDEX = NO
 DOCBOOK_OUTPUT = docbook
 DOT_GRAPH_MAX_NODES = 50
 DOT_IMAGE_FORMAT = svg
-DOT_NUM_THREADS = 1
+DOT_NUM_THREADS = $(READTHEDOCS_THREADS)
```
Comment (ReenigneArcher) [inline doxyconfig-Doxyfile:110]:
```suggestion
DOT_NUM_THREADS = $(DOXYCONFIG_THREADS)
```

File: doxyconfig-Doxyfile [hunk 3]
```diff
@@ -122,7 +122,7 @@ MACRO_EXPANSION = YES
 MAN_OUTPUT = man
 MARKDOWN_ID_STYLE = GITHUB
 MARKDOWN_SUPPORT = YES
-NUM_PROC_THREADS = 1
+NUM_PROC_THREADS = $(READTHEDOCS_THREADS)
```
Comment (ReenigneArcher) [inline doxyconfig-Doxyfile:125]:
```suggestion
NUM_PROC_THREADS = $(DOXYCONFIG_THREADS)
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -92,13 +92,23 @@ add_custom_command(
 file(RELATIVE_PATH DOXYGEN_BUILD_DIR_RELATIVE "${SOURCE_DOCS_DIR}" "${DOXYGEN_BUILD_DIR_CMAKE}")
 message(STATUS "DOXYGEN_BUILD_DIR_RELATIVE: ${DOXYGEN_BUILD_DIR_RELATIVE}")
 
+if(CMAKE_HOST_WIN32)
+    # On Windows, we have to build the documentation using only a single thread to
+    # avoid the build mysteriously taking forever.
+    # See https://github.com/LizardByte/Sunshine/pull/3361#issuecomment-2453549801.
```
Comment (ReenigneArcher) [inline CMakeLists.txt:98]:
```suggestion
    # See https://github.com/doxygen/doxygen/issues/9694
```

Probably better to just link to the upstream issue?

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
Thank you again! Just a couple of suggestions below.

I don't remember if there's a way to set a default value if the env value is empty or not. If NOT the env value also needs to be set before we run this command.

https://github.com/LizardByte/doxyconfig/blob/4c051982c1523f88fbbdb69d44af2ea384dc7052/readthedocs_build.sh#L60

The build in readthedocs does not use cmake, since it would require installing all the project dependencies and such, wasting a lot of time.

**Comment by cathyjf** [comment]
> I don't remember if there's a way to set a default value if the env value is empty or not. If NOT the env value also needs to be set before we run this command.

I can't find any way to set a default value in the doxygen config file. However, if the environment variable is blank, that doesn't cause a syntax error. Instead, it's equivalent to setting a blank value for the variable. This in turn causes the default value for the variable to be used. This is fine for our case because it means that multithreaded will be enabled by default, unless disabled by the cmake variable. Your `readthedocs` configuration suggests that Ubuntu will be used to generate the docs, which is not Windows, so using multithreading there should be fine.

**Review by ReenigneArcher — APPROVED** [review]

**Pull request merged**